components/Filter.js

```diff
@@ -28,6 +29,22 @@ export default function Filter({
     }
   }
 
+  function getCategoriesWithTransactionCount(transactionsList) {
+    return transactionsList.reduce((acc, transaction) => {
+      const categoryId =
+        transaction.category?._id || transaction.category || "uncategorized";
+      acc[categoryId] = (acc[categoryId] || 0) + 1;
+      return acc;
+    }, {});
+  }
+
+  const categoriesWithCounts = categories.map((category) => {
+    const counts = getCategoriesWithTransactionCount(transactionsList);
+    return {
+      ...category,
+      count: counts[category._id] || 0,
+    };
```

Comment:
you have a small proplem here. you generate the entire list of transaction counts for each category. you call getCategoriesWithTransactionCount for each category, but every function call calculates the count for each category. You can simply create the count and then look the value up:
```suggestion
    const categryCounts = getCategoriesWithTransactionCount(transactions);
    
    const categoriesWithCounts = categories.map((category) => {
    return {
      ...category,
      count: categryCounts[category._id] || 0,
    };
```

Reply:
Thank you for the hint. I had useMemo hook here, but i was told not to use it, thats why i choose this way, but your code makes sense.